tests/src/Controller/Cas20ControllerTest.php

```diff
@@ -0,0 +1,588 @@
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+use SimpleSAML\CAS\Constants as C;
+use SimpleSAML\Configuration;
+use SimpleSAML\Module;
+use SimpleSAML\Module\casserver\Cas\Factories\TicketFactory;
+use SimpleSAML\Module\casserver\Cas\Ticket\FileSystemTicketStore;
+use SimpleSAML\Module\casserver\Cas\TicketValidator;
+use SimpleSAML\Module\casserver\Controller\Cas20Controller;
+use SimpleSAML\Session;
+use SimpleSAML\Utils;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class Cas20ControllerTest extends TestCase
+{
+    private array $moduleConfig;
+
+    private Session $sessionMock;
+
+    private Request $samlValidateRequest;
+
+    private string $sessionId;
+
+    private Configuration $sspConfig;
+
+    private FileSystemTicketStore $ticketStore;
+
+    private TicketValidator $ticketValidatorMock;
+
+    private Utils\HTTP|MockObject $utilsHttpMock;
+
+    private array $ticket;
+
+    /**
+     * @throws \Exception
+     */
+    protected function setUp(): void
+    {
+        $this->sspConfig    = Configuration::getConfig('config.php');
+        $this->sessionId    = session_create_id();
+        $this->moduleConfig = [
+            'ticketstore' => [
+                'class'     => 'casserver:FileSystemTicketStore', //Not intended for production
+                'directory' => __DIR__ . '../../../../tests/ticketcache',
+            ],
+        ];
+
+        // Hard code the ticket store
+        $this->ticketStore = new FileSystemTicketStore(Configuration::loadFromArray($this->moduleConfig));
+
+        $this->ticketValidatorMock = $this->getMockBuilder(TicketValidator::class)
+            ->setConstructorArgs([Configuration::loadFromArray($this->moduleConfig)])
+            ->onlyMethods(['validateAndDeleteTicket'])
+            ->getMock();
+
+        $this->sessionMock = $this->getMockBuilder(Session::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['getSessionId'])
+            ->getMock();
+
+        $this->utilsHttpMock = $this->getMockBuilder(Utils\HTTP::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['fetch'])
+            ->getMock();
+
+        $this->ticket = [
+            'id'          => 'ST-' . $this->sessionId,
+            'validBefore' => 1731111111,
+            'service'     => 'https://myservice.com/abcd',
+            'forceAuthn'  => false,
+            'userName'    => 'username@google.com',
+            'attributes'  =>
+                [
+                    'eduPersonPrincipalName' =>
+                        [
+                            0 => 'eduPersonPrincipalName@google.com',
+                        ],
+                ],
+            'proxies'     =>
+                [
+                ],
+            'sessionId'   => $this->sessionId,
+        ];
+    }
+
+    public function testProxyValidatePassesTheCorrectMethodToValidate(): void
```

Comment:
ah, I think this should be changed to actually confirm that the right thing happens. Otherwise we don't have any coverage for proxy validate functionality.